### PR TITLE
support running workloads in visibility endpoint

### LINKED ISSUE #1776: A mechanism to list the Workloads that are admitted and not finished
**What would you like to be added**:

Some way of querying the list of Workloads that are still "running", that is, they are admitted and don't have a Finished condition. Querying by ClusterQueue would be ideal too.

Some options (not necessarily exclusive to each other):
- The kueue CLI
- A visibility endpoint.
- Is there a way to do this just with a kubectl jsonpath?

**Why is this needed**:

For basic diagnostics of the state of the cluster.

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change?
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by KunWuLuan**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

support running workloads in visibility endpoint

#### Which issue(s) this PR fixes:

Fixes #1776

#### Does this PR introduce a user-facing change?

```release-note
We will add a new visibility endpoint in Kueue. Users can list the running workloads by using 
'kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/runningworkloads"' 
and 
'kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/namespaces/default/localqueues/user-queue/runningworkloads"'
```

### CODE REVIEW

**Comment by KunWuLuan** [comment]
do we need to query running workloads for local queues?

**Review by PBundyra — CHANGES_REQUESTED** [review]
Thanks for the PR!

My first pass.

Please add e2e/integration tests and update KEP regarding that. Please add cases which cover:
- pending and running workloads are assigned to a CQ
- a workload changes its status from pending to admitted
- a workload changes its status from admitted to finished

**Comment by PBundyra** [comment]
> do we need to query running workloads for local queues?

I am leaning towards implementing that. It will be useful for the KueueCtl 
@alculquicondor @mwielgus WDYT?

File: pkg/visibility/api/rest/running_workloads_cq_test.go
Comment (PBundyra) [inline pkg/visibility/api/rest/running_workloads_cq_test.go:1]: Please add cases which cover pending and finished workloads

File: pkg/cache/cache.go [hunk 1]
```diff
@@ -657,6 +661,41 @@ func filterLocalQueueUsage(orig FlavorResourceQuantities, resourceGroups []Resou
...
+	defer c.RUnlock()
+	cq, ok := c.clusterQueues[name]
+	if !ok {
+		return []*workload.Info{}, apierr.NewNotFound(v1alpha1.Resource("clusterqueue"), name)
+	}
+
+	count := 0
+	wkls := make([]*workload.Info, 0, len(cq.Workloads))
+	for k, wkl := range cq.Workloads {
+		if _, ok := c.assumedWorkloads[k]; ok {
```
Comment (PBundyra) [inline pkg/cache/cache.go:686]: I am not sure about this condition as well, could you clarify the intentions please? Another way of checking if a workload is running would be checking its conditions - it's accessible in `wkl` object here

**Comment by alculquicondor** [comment]
> > do we need to query running workloads for local queues?
> 
> I am leaning towards implementing that. It will be useful for the KueueCtl. It might be a separate PR though

Yes, a separate PR would be better.

File: apis/visibility/v1alpha1/types.go [hunk 2]
```diff
@@ -77,6 +79,35 @@ type PendingWorkload struct {
 	PositionInLocalQueue int32 `json:"positionInLocalQueue"`
 }
 
+// RunningWorkload is a user-facing representation of a running workload that summarizes the relevant information for
+// assumed resources in the cluster queue.
+type RunningWorkload struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Priority indicates the workload's priority
+	Priority int32 `json:"priority"`
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply apis/visibility/v1alpha1/types.go:88]: how about add a field that shows the time the workloads are admited?

File: pkg/visibility/api/rest/running_workloads_cq.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+var (
+	debugLog = ctrl.Log.WithName("visibility-server")
+)
+
+// Get implements rest.GetterWithOptions interface
+// It fetches information about pending workloads and returns according to query params
+func (m *runningWorkloadsInCqREST) Get(ctx context.Context, name string, opts runtime.Object) (runtime.Object, error) {
+	debugLog.V(1).Info("got a request")
+	pendingWorkloadOpts, ok := opts.(*v1alpha1.PendingWorkloadOptions)
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq.go:65]: sure

File: pkg/visibility/api/rest/running_workloads_cq_test.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+
+					if p1 != p2 {
+						return p1 > p2
+					}
+
+					return a.CreationTimestamp.Before(&b.CreationTimestamp)
+				}
+				sort.Slice(tc.wantResp.wantRunningWorkloads, func(i, j int) bool {
+					return less(tc.wantResp.wantRunningWorkloads[i], tc.wantResp.wantRunningWorkloads[j])
+				})
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq_test.go:427]: I used the same test case as the previous one, but the sorting in the previous case does not work for the new use case. For convenience, and to facilitate future modifications to the test case, I decided to add a sorting process here.

File: pkg/visibility/api/rest/running_workloads_cq_test.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/constants"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+)
+
+func TestRunningWorkloadsInCQ(t *testing.T) {
+	const (
+		nsName     = "foo"
+		cqNameA    = "cqA"
+		cqNameB    = "cqB"
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq_test.go:42]: deleted the constant cqNameB

File: pkg/visibility/api/rest/running_workloads_cq_test.go [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+		lqNameA    = "lqA"
+		lqNameB    = "lqB"
+		flavorName = "flavor"
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq_test.go:45]: deleted the constant flavorName

File: pkg/visibility/api/rest/running_workloads_cq_test.go [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+				utiltesting.MakeLocalQueue(lqNameA, nsName).ClusterQueue(cqNameA).Obj(),
+			},
+			workloads: []*kueue.Workload{
+				utiltesting.MakeWorkload("a", nsName).PodSets(podSets...).ReserveQuota(&kueue.Admission{
+					ClusterQueue:      cqNameA,
+					PodSetAssignments: podSetFlavors,
+				}).Queue(lqNameA).Priority(highPrio).Creation(now).Admission(adA).Obj(),
+				utiltesting.MakeWorkload("b", nsName).PodSets(podSets...).ReserveQuota(&kueue.Admission{
+					ClusterQueue:      cqNameA,
+					PodSetAssignments: podSetFlavors,
```
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq_test.go:118]: done

File: apis/visibility/v1alpha1/types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (PBundyra) [reply apis/visibility/v1alpha1/types.go:88]: Sounds good

File: pkg/visibility/api/rest/running_workloads_cq.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+func (m *runningWorkloadsInCqREST) Destroy() {}
+
+var (
+	debugLog = ctrl.Log.WithName("visibility-server")
+)
+
+// Get implements rest.GetterWithOptions interface
+// It fetches information about running workloads and returns according to query params
+func (m *runningWorkloadsInCqREST) Get(ctx context.Context, name string, opts runtime.Object) (runtime.Object, error) {
+	debugLog.V(1).Info("got a request")
```
Comment (PBundyra) [inline pkg/visibility/api/rest/running_workloads_cq.go:64]: Please change logger verbosity to 2

File: pkg/visibility/api/rest/running_workloads_cq.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	runningWorkloadOpts, ok := opts.(*v1alpha1.RunningWorkloadOptions)
+	if !ok {
+		return nil, fmt.Errorf("invalid options object: %#v", opts)
+	}
+	limit := runningWorkloadOpts.Limit
+	offset := runningWorkloadOpts.Offset
+
+	wls := make([]v1alpha1.RunningWorkload, 0, limit)
+	runningWorkloadsInfo, err := m.c.RunningWorkload(name)
+	debugLog.V(1).Info("got runningWorkloads", "len", len(runningWorkloadsInfo))
```
Comment (PBundyra) [inline pkg/visibility/api/rest/running_workloads_cq.go:74]: ditto

File: pkg/visibility/api/rest/running_workloads_cq.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	for index := int(offset); index < int(offset+limit) && index < len(runningWorkloadsInfo); index++ {
+		wlInfo := runningWorkloadsInfo[index]
+		wls = append(wls, *newRunningWorkload(wlInfo))
+	}
+	debugLog.V(1).Info("response", "len", len(wls))
```
Comment (PBundyra) [inline pkg/visibility/api/rest/running_workloads_cq.go:83]: ditto

File: pkg/visibility/api/rest/running_workloads_cq.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+func (m *runningWorkloadsInCqREST) New() runtime.Object {
+	return &v1alpha1.RunningWorkloadsSummary{}
+}
+
+// Destroy implements rest.Storage interface
+func (m *runningWorkloadsInCqREST) Destroy() {}
+
+var (
+	debugLog = ctrl.Log.WithName("visibility-server")
+)
```
Comment (PBundyra) [inline pkg/visibility/api/rest/running_workloads_cq.go:59]: Could we have one logger in this file? Use this one and delete the logger from the `runningWorkloadsInCqREST` struct or the other way around please. If you decide to stick with this one, please rename it to `log` for consitency

**Comment by PBundyra** [comment]
Thank you for addressing my comments. Please add missing test scenarios.

File: pkg/visibility/api/rest/running_workloads_cq_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq_test.go:335]: test cases in pending_workloads_cq_test.go

File: pkg/visibility/api/rest/running_workloads_cq.go (hunk 11 above)
Comment (KunWuLuan) [reply pkg/visibility/api/rest/running_workloads_cq.go:59]: This variable was used by me during the feature development, and I have removed it as it's no longer necessary.

File: keps/2145-list-admitted-workloads/README.md
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:1]: Please also fill information about new API objects

**Comment by KunWuLuan** [comment]
@PBundyra Hi, I think all questions have been addressed. Please have a look if you have time. Thanks very much.

File: keps/2145-list-admitted-workloads/README.md [hunk 12]
```diff
@@ -83,7 +128,7 @@ implementing this enhancement to ensure the enhancements have also solid foundat
 
 #### Unit Tests
 
-New unit tests should be added testing the functionality for jobs and pods.
+New unit tests should be added testing the functionality for new api.
 
 #### Integration tests
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:134]: Please update this section

File: keps/2145-list-admitted-workloads/README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+Add a new visibility endpoint in Kueue for querying the list of Workloads that are still "running".
+
+## Motivation
+
+Jsonpath support in kubectl is limited, and we can not filter resources by condition in this way. By adding a new
+visibility endpoint, users can list running workloads by `kubectl get`.
+
+### Goals
+
+* Support list workloads that are admitted and not finished.
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:32]:
```suggestion
* Support listing running workloads in a ClusterQueue
```

File: keps/2145-list-admitted-workloads/README.md
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:1]: Could you also add information about the LocalQueue endpoint, even if it won't be implemented immediately please.

Please add reference to [this KEP](https://github.com/kubernetes-sigs/kueue/tree/main/keps/168-2-pending-workloads-visibility)

File: keps/2145-list-admitted-workloads/README.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+	Limit int64 `json:"limit,omitempty"`
+}
+```
+
+Users can list the running workloads by using
+``` bash
+kubectl get --raw "/apis/visibility.kueue.x-k8s.io/v1alpha1/clusterqueues/cluster-queue/runningworkloads"
+```
+
+We will show priority and localqueue information in response. Like this:
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:93]: Please update

File: keps/2145-list-admitted-workloads/README.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+```
+{"kind":"RunningWorkloadsSummary","apiVersion":"visibility.kueue.x-k8s.io/v1alpha1","metadata":{"creationTimestamp":null},"items":[{"metadata":{"name":"job-sample-job-jz228-ef938","namespace":"default","creationTimestamp":"2024-05-06T02:15:26Z","ownerReferences":[{"apiVersion":"batch/v1","kind":"Job","name":"sample-job-jz228","uid":"2de8a359-4c95-4159-b677-0279066149b6"}]},"priority":0,"admissionTime":"xxxx"}]}
+```
+
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/README.md:97]: Please move this section to Design Details, and add a REST path to the endpoint (not only kubectl usage)

File: keps/2145-list-admitted-workloads/kep.yaml [hunk 16]
```diff
@@ -0,0 +1,22 @@
+title: List Admitted And Not Finished Workloads
+kep-number: 1834
+authors:
+  - "@kunwuluan"
+status: draft
+creation-date: 2024-04-29
+reviewers:
+  - "@alculquicondor"
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/kep.yaml:8]:
```suggestion
  - "@alculquicondor"
  - "@pbundyra"
```

File: keps/2145-list-admitted-workloads/kep.yaml [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v0.9"
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v0.9"
+  beta: "v0.9"
```
Comment (PBundyra) [inline keps/2145-list-admitted-workloads/kep.yaml:22]: Please update

**Comment by PBundyra** [comment]
> @PBundyra Hi, I think all questions have been addressed. Please have a look if you have time. Thanks very much.

Thank you for addressing my comments. Please add e2e/integration tests as I've mentioned above - you can extend [existing e2e tests](https://github.com/kubernetes-sigs/kueue/blob/main/test/e2e/singlecluster/visibility_test.go).

Another thing we're missing here are RBAC roles. See [this file](https://github.com/kubernetes-sigs/kueue/commit/ec340f28f509ceab705ee9fe998e19e8b729bd50#diff-05b522ad25b55cd648d0bb47c0686d7a8d97c1c19b9b533010f8750b23199b0c) for the reference

File: keps/2145-list-admitted-workloads/README.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+<!-- /toc -->
+
+## Summary
+
+Add a new visibility endpoint in Kueue for querying the list of Workloads that are still "running".
+
+## Motivation
+
+Jsonpath support in kubectl is limited, and we can not filter resources by condition in this way. By adding a new
+visibility endpoint, users can list running workloads by `kubectl get`.
```
Comment (mwielgus) [inline keps/2145-list-admitted-workloads/README.md:28]: This functionality is already being worked on in kueuectl list workloads (see: https://github.com/kubernetes-sigs/kueue/blob/8d7f709225bb828c38758a22ccf892f304b1bd0c/keps/2076-kueuectl/README.md#list-workloads 
and 
https://github.com/kubernetes-sigs/kueue/pull/2195). Is there another reason why we might need this extra api?

File: keps/2145-list-admitted-workloads/README.md
Comment (KunWuLuan) [reply keps/2145-list-admitted-workloads/README.md:1]: Sure

File: keps/2145-list-admitted-workloads/README.md (hunk 18 above)
Comment (PBundyra) [reply keps/2145-list-admitted-workloads/README.md:28]: @KunWuLuan before we go further with this PR could you please address that?

**Pull request closed without merging**